sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -180,37 +185,49 @@ private[sbtplugin] object ScalaJSPluginInternal {
 
-          FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
-              FilesInfo.exists) { _ => // We don't need the files
+          def moduleInitializersChanged = (scalaJSModuleInitializersFingerprints in key)
+            .previous
+            .exists(_ != (scalaJSModuleInitializersFingerprints in key).value)
 
-            val stageName = stage match {
-              case Stage.FastOpt => "Fast"
-              case Stage.FullOpt => "Full"
-            }
+          def linkerConfigChanged = (scalaJSLinkerConfigFingerprint in key)
+            .previous
+            .exists(_ != (scalaJSLinkerConfigFingerprint in key).value)
+
+          val configChanged = moduleInitializersChanged || linkerConfigChanged
+          if (configChanged && output.exists()) {
+            output.delete() // triggers re-linking through FileFunction.cached
+          }
 
-            log.info(s"$stageName optimizing $output")
+          FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
+            FilesInfo.exists) { _ => // we don't need the files
+              val stageName = stage match {
+                case Stage.FastOpt => "Fast"
+                case Stage.FullOpt => "Full"
+              }
 
-            IO.createDirectory(output.getParentFile)
+              log.info(s"$stageName optimizing $output")
 
-            def relURI(path: String) = new URI(null, null, path, null)
+              IO.createDirectory(output.getParentFile)
 
-            val out = LinkerOutput(linkerImpl.outputFile(output.toPath))
-              .withSourceMap(linkerImpl.outputFile(sourceMapFile.toPath))
-              .withSourceMapURI(relURI(sourceMapFile.getName))
-              .withJSFileURI(relURI(output.getName))
+              def relURI(path: String) = new URI(null, null, path, null)
 
-            try {
-              enhanceIRVersionNotSupportedException {
-                val tlog = sbtLogger2ToolsLogger(log)
-                await(log)(linker.link(ir, moduleInitializers, out, tlog)(_))
+              val out = LinkerOutput(linkerImpl.outputFile(output.toPath))
+                .withSourceMap(linkerImpl.outputFile(sourceMapFile.toPath))
+                .withSourceMapURI(relURI(sourceMapFile.getName))
+                .withJSFileURI(relURI(output.getName))
+
+              try {
+                enhanceIRVersionNotSupportedException {
+                  val tlog = sbtLogger2ToolsLogger(log)
+                  await(log)(linker.link(ir, moduleInitializers, out, tlog)(_))
+                }
+              } catch {
+                case e: LinkingException =>
+                  throw new MessageOnlyException(e.getMessage)
               }
-            } catch {
-              case e: LinkingException =>
-                throw new MessageOnlyException(e.getMessage)
-            }
 
-            Set(output, sourceMapFile)
-          } (realFiles.toSet)
+              Set(output, sourceMapFile)
+            } (realFiles.toSet)
```

Comment:
I think the indentation is all broken here, making the diff much larger than it actually is.

Reply:
True, my bad. Fixed in 45db847395e8e303891e73da7fc50aaf4b15c930